usb: stm32: added usb_dc_detach code

StefJar opened the pull request:
The usb device is now deinit. As well the
usb endpoints are getting closed.

After deinit the usb periphal wil go to deepsleep.
That might save some power.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

Commit:
usb: stm32: added usb_dc_detach code

The usb device is now deinit. As well the
usb endpoints are getting closed.

After deinit the usb periphal wil go to deepsleep.
That might save some power.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

StefJar commented:
@jfischer-phytec-iot pls review

StefJar commented:
looks like that one test is broken:
```
===================================================================
starting test - test_k_sleep
 thread sleeping for 50 milliseconds
 thread back from sleep
FAIL - thread_sleep@707. timestamp out of range, got 53

    Assertion failed at ZEPHYR_BASE/tests/kernel/context/src/main.c:775: test_k_sleep: (rv not equal to 0)
 *** thread timed out waiting for thread on k_sleep().
FAIL - test_k_sleep
```

looks like the `k_sleep` test fails

ydamigos commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -846,7 +846,9 @@ int usb_dc_ep_mps(const u8_t ep)
-	LOG_ERR("Not implemented");
+	HAL_PCD_DeInit(&usb_dc_stm32_state.pcd);
+	HAL_PCD_EP_Close(&usb_dc_stm32_state.pcd, EP0_IN);
+	HAL_PCD_EP_Close(&usb_dc_stm32_state.pcd, EP0_OUT);
```
According to `usb_dc_detach()`, I am expecting the function to disable clock, interrupt and deinit USB device.

Could you try to remove code closing EP0 and add code disabling clock and interrupts and check if it works in your use case?
Does USB in your application attach again after it detaches?

StefJar replied:
@ydamigos usb re-attach:
Good point, I was asking my self about that as well. But I can't test it, because I set all peripherals off and go to sleep mode in my current firmware. 

at clock disable:
think we need a similar function like the `usb_dc_stm32_clock_enable`. If you are looking into it you will see alot of #defines for the different stm32 arch. I even don't have a eval board here. So I can provide only some working code for the stm32f412. Think this would be a perfect task for the st guys(@erwango)
IMHO I think the perfect way would looks like: usb off, endpoints off, ISR off, clk off, pins off
I will add `usb_dc_stm32_clock_disable` and do a clk off there. This code will be quite basic, so the users need to add the clk off code for other arch than stm32f412

at isr:
I will add `irq_disable(DT_USB_IRQ)` no prob good point.

at remove endpoints:
For me it comes to the question: When the usb is gone why the endpoints are alive? My intention is to put them to a defined state. The close state seems to be a best one. But off course I am open to other ideas.

ydamigos replied:
>at remove endpoints:
For me it comes to the question: When the usb is gone why the endpoints are alive? My intention is to put them to a defined state. The close state seems to be a best one. But off course I am open to other ideas.

I was expecting the application to call `usb_dc_ep_disable` to close all endpoints before calling `usb_disable`.

StefJar replied:
> > at remove endpoints:
> > For me it comes to the question: When the usb is gone why the endpoints are alive? My intention is to put them to a defined state. The close state seems to be a best one. But off course I am open to other ideas.
> 
> I was expecting the application to call `usb_dc_ep_disable` to close all endpoints before calling `usb_disable`.

I am using the usb as CDC ACM `CDC_ACM_0`. 
As far as I remember I took the isr driven example code(/samples/subsys/usb/cdc_acm) as base.

```
	// set up usb
	usbACM.uart = device_get_binding(USBACM_DEV_NAME);
	if (!usbACM.uart) {
		LOG_ERR("CDC ACM \"" USBACM_DEV_NAME "\" device not found");
		return -1;
	}

	// enable uart data in isr
	uart_irq_callback_set(usbACM.uart, usbACM_isr);

	// enable pins
	gpio_pins_usb(true);

	if (0 != usb_enable(usbACM_status_cb)) {
		LOG_ERR("Failed to enable USB");
		return -1;
	}

	// set the acm baudrate
	ret = uart_line_ctrl_get(usbACM.uart, LINE_CTRL_BAUD_RATE, &baudrate);
	if (ret) {
		LOG_ERR("Failed to get baudrate, ret code %d", ret);
		return -2;
	} else {
		LOG_DBG("Baudrate detected: %d", baudrate);
	}
	// try to set baudrate
	baudrate = USB_ACM_BAUDRATE;
	LOG_DBG("try to set baudrate to %d", baudrate);
	ret = uart_line_ctrl_set(usbACM.uart, LINE_CTRL_BAUD_RATE, baudrate);
	if (ret) {
		LOG_ERR("Failed to set baudrate to %d, ret code %d", baudrate, ret);
		return -3;
	} else {
		uart_line_ctrl_get(usbACM.uart, LINE_CTRL_BAUD_RATE, &baudrate);
		LOG_DBG("Baudrate set to %d", baudrate);
	}

	// enable rx isr
	uart_irq_rx_enable(usbACM.uart);

```
As you can see and check, there is no interaction to the endpoints from the user/app-space. As app I have no API to read the endpoints back (for the usb cdc acm). That is why I was thinking about closing them.

I rechecked "usb_disable()" implementation. It is going to "usb_disable" @ "usb_device.c" and then directly into "usb_dc_stm32.c". So no endpoint handling at this code path.

ydamigos replied:
I see your point but all the other drivers don't handle endpoints in `usb_dc_detach()`.

StefJar replied:
I updated the commit. So it does not handle the endpoints.
As well I merged my stuff with the latest changes on master.

After discussing with @ydamigos, I will open an bug because the stm32 usb driver is not closing his endpoints. This will happen when we are through the commit process.

Commit:
usb: stm32: added stop clk and disable interrupt

usb_dc_detach now stops the clk and disables the
interrupt.

The clk disable is only doing a std. clk disable.
No arch spezific deinit code is done. That my be
added later.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

Commit:
usb: stm32: removed empty line that breaks checkpatch

Removed empty line that breaks checkpatch.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

ydamigos commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -264,6 +264,16 @@ static int usb_dc_stm32_clock_enable(void)
+	};
+	clock_control_off(clk, (clock_control_subsys_t *)&pclken);
+}
+
```
Check PR #16026. It creates a `static struct stm32_pclken pclken` and uses it for both  `clock_control_on` and `clock_control_off`.

Commit:
usb: stm32: removed endpoint deinit

merged with current master and removed the endpoint
handling in the detach function.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

erwango commented:
@StefJar do you plan to continue this PR?

StefJar commented:
hey ho
I am happy if we can roll in this changes.
if I get it right @ydamigos wants me to add a `usb_dc_ep_disable` at the close function?
